File: core/src/main/java/org/evergreen_ils/auth/AccountAuthenticator.java
package org.evergreen_ils.auth;

import android.accounts.AbstractAccountAuthenticator;
import android.accounts.Account;
import android.accounts.AccountAuthenticatorResponse;
import android.accounts.AccountManager;
import android.accounts.NetworkErrorException;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import org.evergreen_ils.R;
import org.evergreen_ils.android.Analytics;

public class AccountAuthenticator extends AbstractAccountAuthenticator {
    
    private final static String TAG = AccountAuthenticator.class.getSimpleName();
    private Context context;
    private Class authenticatorActivity;

    public AccountAuthenticator(Context context) {
        super(context);
        this.context = context;

        // Choose the right AuthenticatorActivity.  A custom app does not require the library spinner.
        String library_url = context.getString(R.string.ou_library_url);
        if (TextUtils.isEmpty(library_url)) {
            this.authenticatorActivity = GenericAuthenticatorActivity.class;
        } else {
            this.authenticatorActivity = AuthenticatorActivity.class;
        }
    }

    @Override
    public Bundle addAccount(AccountAuthenticatorResponse response, String accountType, String authTokenType, String[] requiredFeatures, Bundle options) throws NetworkErrorException {
        Analytics.log(TAG, "addaccount "+accountType+" "+authTokenType);
        final Intent intent = new Intent(context, authenticatorActivity);
        // setting ARG_IS_ADDING_NEW_ACCOUNT here does not work, because this is not the
        // same Intent as the one in AuthenticatorActivity.finishLogin
        //intent.putExtra(AuthenticatorActivity.ARG_IS_ADDING_NEW_ACCOUNT, true);
        intent.putExtra(AccountManager.KEY_ACCOUNT_AUTHENTICATOR_RESPONSE, response);
        
        Bundle result = new Bundle();
        result.putParcelable(AccountManager.KEY_INTENT, intent);
        return result;
    }

    @Override
    public Bundle getAuthToken(AccountAuthenticatorResponse response, Account account, String authTokenType, Bundle options) throws NetworkErrorException {
        Analytics.log(TAG, "getAuthToken> "+account.name);

        // If the caller requested an authToken type we don't support, then
        // return an error
        if (!authTokenType.equals(Const.AUTHTOKEN_TYPE)) {
            final Bundle result = new Bundle();
            result.putString(AccountManager.KEY_ERROR_MESSAGE, "invalid authTokenType");
            return result;
        }

        final AccountManager am = AccountManager.get(context);
        String library_name = am.getUserData(account, Const.KEY_LIBRARY_NAME);
        String library_url = am.getUserData(account, Const.KEY_LIBRARY_URL);
        Analytics.log(TAG, "getAuthToken> library_name=" + library_name + " library_url=" + library_url);
        if (library_name == null) {
            // workaround issue #24 - not sure how it happened
            library_name = context.getString(R.string.ou_library_name);
        }
        if (library_url == null) {
            // workaround issue #24 - not sure how it happened
            library_url = context.getString(R.string.ou_library_url);
        }

        String authToken = am.peekAuthToken(account, authTokenType);
        Analytics.log(TAG, "getAuthToken> peekAuthToken returned " + authToken);
        if (TextUtils.isEmpty(authToken)) {
            final String password = am.getPassword(account);
            if (password != null) {
                try {
                    Analytics.log(TAG, "getAuthToken> attempting to sign in with existing password");
                    authToken = EvergreenAuthenticator.signIn(library_url, account.name, password);
                } catch (AuthenticationException e) {
                    Analytics.logException(e);
                    am.clearPassword(account);
                    final Bundle result = new Bundle();
                    result.putString(AccountManager.KEY_ERROR_MESSAGE, e.getMessage());
                    return result;
                } catch (Exception e2) {
                    Analytics.logException(e2);
                    am.clearPassword(account);
                    final Bundle result = new Bundle();
                    result.putString(AccountManager.KEY_ERROR_MESSAGE, "Sign in failed");
                    return result;
                }
            }
        }

        // If we get an authToken - we return it
        Analytics.log(TAG, "getAuthToken> token "+Analytics.redactedString(authToken));
        if (!TextUtils.isEmpty(authToken)) {
            final Bundle result = new Bundle();
            result.putString(AccountManager.KEY_ACCOUNT_NAME, account.name);
            result.putString(AccountManager.KEY_ACCOUNT_TYPE, account.type);
            result.putString(AccountManager.KEY_AUTHTOKEN, authToken);
            result.putString(Const.KEY_LIBRARY_NAME, library_name);
            result.putString(Const.KEY_LIBRARY_URL, library_url);
            return result;
        }

        // If we get here, then we couldn't access the user's password - so we
        // need to re-prompt them for their credentials. We do that by creating
        // an intent to display our AuthenticatorActivity.
        Analytics.log(TAG, "getAuthToken> creating intent to display AuthenticatorActivity");
        final Intent intent = new Intent(context, authenticatorActivity);
        intent.putExtra(AccountManager.KEY_ACCOUNT_AUTHENTICATOR_RESPONSE, response);
        intent.putExtra(AuthenticatorActivity.ARG_ACCOUNT_TYPE, account.type);
        intent.putExtra(AuthenticatorActivity.ARG_AUTH_TYPE, authTokenType);
        intent.putExtra(AuthenticatorActivity.ARG_ACCOUNT_NAME, account.name);
        final Bundle bundle = new Bundle();
        bundle.putParcelable(AccountManager.KEY_INTENT, intent);
        return bundle;
    }

    @Override
    public String getAuthTokenLabel(String authTokenType) {
        return Const.AUTHTOKEN_TYPE_LABEL;
    }

    @Override
    public Bundle hasFeatures(AccountAuthenticatorResponse response, Account account, String[] features) throws NetworkErrorException {
        Analytics.log(TAG, "hasFeatures "+account.name+" features "+features);
        final Bundle result = new Bundle();
        result.putBoolean(AccountManager.KEY_BOOLEAN_RESULT, false);
        return result;
    }

    @Override
    public Bundle editProperties(AccountAuthenticatorResponse response, String accountType) {
        Analytics.log(TAG, "editProperties "+accountType);
        return null;
    }

    @Override
    public Bundle confirmCredentials(AccountAuthenticatorResponse response, Account account, Bundle options) throws NetworkErrorException {
        Analytics.log(TAG, "confirmCredentials "+account.name);
        return null;
    }

    @Override
    public Bundle updateCredentials(AccountAuthenticatorResponse response, Account account, String authTokenType, Bundle options) throws NetworkErrorException {
        Analytics.log(TAG, "updateCredentials "+account.name);
        return null;
    }
}

File: core/src/main/java/net/kenstir/ui/account/AccountAuthenticator.kt
/*
 * Copyright (c) 2025 Kenneth H. Cox
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <https://www.gnu.org/licenses/>.
 */
package net.kenstir.ui.account

import android.accounts.AbstractAccountAuthenticator
import android.accounts.Account
import android.accounts.AccountAuthenticatorResponse
import android.accounts.AccountManager
import android.accounts.NetworkErrorException
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import kotlinx.coroutines.runBlocking
import net.kenstir.hemlock.R
import net.kenstir.util.Analytics.log
import net.kenstir.util.Analytics.redactedString
import net.kenstir.ui.App
import org.evergreen_ils.gateway.GatewayClient

class AccountAuthenticator(private val context: Context): AbstractAccountAuthenticator(context) {
    private var authenticatorActivity: Class<*>? = null

    init {
        // Choose the right AuthenticatorActivity.  A custom app does not require the library spinner.
        val libraryUrl = context.getString(R.string.ou_library_url)
        if (TextUtils.isEmpty(libraryUrl)) {
            this.authenticatorActivity = GenericAuthenticatorActivity::class.java
        } else {
            this.authenticatorActivity = AuthenticatorActivity::class.java
        }
    }

    @Throws(NetworkErrorException::class)
    override fun addAccount(
        response: AccountAuthenticatorResponse,
        accountType: String,
        authTokenType: String?,
        requiredFeatures: Array<String>?,
        options: Bundle
    ): Bundle? {
        log(TAG, "addaccount $accountType $authTokenType")
        val intent = Intent(context, authenticatorActivity)
        // setting ARG_IS_ADDING_NEW_ACCOUNT here does not work, because this is not the
        // same Intent as the one in AuthenticatorActivity.finishLogin
        //intent.putExtra(AuthenticatorActivity.ARG_IS_ADDING_NEW_ACCOUNT, true);
        intent.putExtra(AccountManager.KEY_ACCOUNT_AUTHENTICATOR_RESPONSE, response)

        val result = Bundle()
        result.putParcelable(AccountManager.KEY_INTENT, intent)
        return result
    }

    @Throws(NetworkErrorException::class)
    override fun getAuthToken(
        response: AccountAuthenticatorResponse,
        account: Account,
        authTokenType: String,
        options: Bundle
    ): Bundle {
        log(TAG, "getAuthToken> " + redactedString(account.name))

        // If the caller requested an authToken type we don't support, then
        // return an error
        if (authTokenType != Const.AUTHTOKEN_TYPE) {
            val result = Bundle()
            result.putString(AccountManager.KEY_ERROR_MESSAGE, "invalid authTokenType")
            return result
        }

        val am = AccountManager.get(context)
        var libraryName = am.getUserData(account, Const.KEY_LIBRARY_NAME)
        var libraryUrl = am.getUserData(account, Const.KEY_LIBRARY_URL)
        log(TAG, "getAuthToken> library_name=$libraryName library_url=$libraryUrl")
        if (libraryName == null) {
            // workaround issue #24 - not sure how it happened
            libraryName = context.getString(R.string.ou_library_name)
        }
        if (libraryUrl == null) {
            // workaround issue #24 - not sure how it happened
            libraryUrl = context.getString(R.string.ou_library_url)
        }

        var authToken = am.peekAuthToken(account, authTokenType)
        log(TAG, "getAuthToken> peekAuthToken returned " + redactedString(authToken))
        if (TextUtils.isEmpty(authToken)) {
            val password = am.getPassword(account)
            if (password != null) {
                try {
                    log(TAG, "getAuthToken> attempting to sign in with existing password")
                    if (libraryUrl != GatewayClient.baseUrl) {
                        // The app changed the server URL; don't send the old password to the new server.
                        // This path clears the password and prompts the user to sign in again.
                        throw AuthenticationException("Server URL changed, please sign in again")
                    }
                    authToken = runBlocking {
                        App.getServiceConfig().authService.getAuthToken(account.name, password).get()
                    }
                } catch (e: AuthenticationException) {
                    //Analytics.logException(e);
                    am.clearPassword(account)
                    val result = Bundle()
                    result.putString(AccountManager.KEY_ERROR_MESSAGE, e.message)
                    return result
                } catch (e2: Exception) {
                    //Analytics.logException(e2);
                    am.clearPassword(account)
                    val result = Bundle()
                    result.putString(AccountManager.KEY_ERROR_MESSAGE, "Sign in failed")
                    return result
                }
            }
        }

        // If we get an authToken - we return it
        log(TAG, "getAuthToken> token " + redactedString(authToken))
        if (!TextUtils.isEmpty(authToken)) {
            val result = Bundle()
            result.putString(AccountManager.KEY_ACCOUNT_NAME, account.name)
            result.putString(AccountManager.KEY_ACCOUNT_TYPE, account.type)
            result.putString(AccountManager.KEY_AUTHTOKEN, authToken)
            result.putString(Const.KEY_LIBRARY_NAME, libraryName)
            result.putString(Const.KEY_LIBRARY_URL, libraryUrl)
            return result
        }

        // If we get here, then we couldn't access the user's password - so we
        // need to re-prompt them for their credentials. We do that by creating
        // an intent to display our AuthenticatorActivity.
        log(TAG, "getAuthToken> creating intent to display AuthenticatorActivity")
        val intent = Intent(context, authenticatorActivity)
        intent.putExtra(AccountManager.KEY_ACCOUNT_AUTHENTICATOR_RESPONSE, response)
        intent.putExtra(AuthenticatorActivity.ARG_ACCOUNT_TYPE, account.type)
        intent.putExtra(AuthenticatorActivity.ARG_AUTH_TYPE, authTokenType)
        intent.putExtra(AuthenticatorActivity.ARG_ACCOUNT_NAME, account.name)
        val bundle = Bundle()
        bundle.putParcelable(AccountManager.KEY_INTENT, intent)
        return bundle
    }

    override fun getAuthTokenLabel(authTokenType: String): String {
        return Const.AUTHTOKEN_TYPE_LABEL
    }

    @Throws(NetworkErrorException::class)
    override fun hasFeatures(
        response: AccountAuthenticatorResponse,
        account: Account,
        features: Array<String>
    ): Bundle {
        log(TAG, "hasFeatures features $features")
        val result = Bundle()
        result.putBoolean(AccountManager.KEY_BOOLEAN_RESULT, false)
        return result
    }

    override fun editProperties(response: AccountAuthenticatorResponse, accountType: String): Bundle? {
        log(TAG, "editProperties $accountType")
        return null
    }

    @Throws(NetworkErrorException::class)
    override fun confirmCredentials(
        response: AccountAuthenticatorResponse,
        account: Account,
        options: Bundle
    ): Bundle? {
        log(TAG, "confirmCredentials")
        return null
    }

    @Throws(NetworkErrorException::class)
    override fun updateCredentials(
        response: AccountAuthenticatorResponse,
        account: Account,
        authTokenType: String,
        options: Bundle
    ): Bundle? {
        log(TAG, "updateCredentials")
        return null
    }

    companion object {
        private const val TAG = "AccountAuthenticator"
    }
}
